### DISCUSSION TITLE
API Engineers

### DISCUSSION THREAD

**Discussion by arneschreuder** [COS-301/graduates#10] [General]
This is the discussion board for and with API Engineers.

**Comment by arneschreuder** [comment]
Please use: https://docs.nestjs.com/graphql/quick-start

**Reply by Grobbies26** [reply]
Should we run the "npm i ..." for apollo and express, or are those not relevant to our project and we shouldnt concern ourselves with that, and just use this as a refernece for graphql

**Reply by arneschreuder** [reply]
So most of the dependencies have already been added apart from GraphQL. I can maybe if there is time tomorrow use GraphQL as an example for the PR. But go check `package.json` to see what has already been added. Unfortunately I did not get tot Apollo and GraphQL so I was not able to add that before.

Side note: The repo uses `yarn`, not npm. You can deduce this from the fact that there is a yarn.lock in the repo.

**Reply by arneschreuder** [reply]
You can also see this from the readme. Most of the scripts show there show yarn.

**Comment by MichaelDuPlessis** [comment]
I assume all code for any api's should go in apps/api? Can anyone confirm.

**Reply by arneschreuder** [reply]
No, only code that is linked up should go in there. All the code that actually gets implemented gets put in libs/apps/api in almost all of the cases. This is the convention for micro-libs. That which is in apps/api is just the executable that loads the application. So you will see something like:

```ts
import { ApiShellFeatureModule } from '@graduates/api/shell/feature';
import { Module } from '@nestjs/common';

@Module({
  imports: [ApiShellFeatureModule], // This loads the API shell, which is the shell bundle/module for the entire api.
})
export class AppModule {}
```

**Reply by arneschreuder** [reply]
Please read the technical requirements I posted in the Wiki. I do mention the micro-libs thing there. I also posted links to references that explain the conventions.

So just for clarity, say I need a StudentsProfiles Domain. Then everything related to student profiles that relates to the api, should follow the micro-lib convention and be put in libs/apps/api/student_profiles/(feature|shared|etc). These feature libs are generated using `nx generate @nrwl/nest:library` (use the VSCode plugin Nx Console -> Generate for this)

**Reply by MichaelDuPlessis** [reply]
When using nx to generate a nest library I assume the path would be libs/apps/api/ but I am unsure about the prefix and rootDir options.

**Reply by Ruben-Brits** [reply]
@MichaelDuPlessis  Have a look at https://nx.dev/structure/library-types to see if this does not help ?

**Reply by MichaelDuPlessis** [reply]
Not particularly as I am struggling to find what should all be put in these fields
<img width="430" alt="Screenshot 2022-03-18 at 15 13 51" src="https://user-images.githubusercontent.com/52035435/159009232-c74ae288-57db-4121-add4-7bc020500543.png">

**Reply by arneschreuder** [reply]
@MichaelDuPlessis  What type of library are you trying to generate?

**Reply by MichaelDuPlessis** [reply]
I see my mistake something of mine was configured incorrectly, if anyone needs help I can try and assist them.

**Reply by MichaelDuPlessis** [reply]
@arneschreuder Just to confirm that I am not doing something wrong but "shared" folder for a specific feature isn't auto-generated when I generate a new library feature?

**Reply by arneschreuder** [reply]
No, shared libraries are explicitly created

**Comment by BenjaminOsmers** [comment]
SO the data engineers will use prisma to model the data and manage the database. Then we, the API Engineers should use graphQL to query the database? Should we then follow these guidlines: https://www.prisma.io/graphql#graphql-tabs ? Or am I wrong?

**Reply by arneschreuder** [reply]
So not entirely, Version 2 of prisma removes the direct graphql syntax, however, this is now incorporates in the prisma client itself. So you should not have to worry about graphql directly. 

Version 1 has this notion of prisma being a repository service, where you query that data/repository service using graphql. But this is no longer the case with version 2 :)

**Reply by arneschreuder** [reply]
O yes, to add, yes to the first part. Data Engineers will use prisma to model the entities in the DB.

**Reply by BenjaminOsmers** [reply]
@arneschreuder Does this look correct? I created a resolver using GraphQL and Prisma to query the DB and retrieve all users
![Screenshot 2022-03-15 at 13 30 48](https://user-images.githubusercontent.com/73130428/158368776-3b45c335-b0bd-4193-babd-b381b2917618.png)

**Reply by BenjaminOsmers** [reply]
@arneschreuder Or should we rather use the NestJS controllers without prisma, as below.
![Screenshot 2022-03-15 at 13 36 53](https://user-images.githubusercontent.com/73130428/158369871-d15a472f-3a8b-4694-993d-82e431572c07.png)

**Reply by arneschreuder** [reply]
Hi @BenjaminOsmers so here is the convention:

* API Layer make resolvers for entities. These are .resolver.ts files. These resolvers inject a service-client and call the...
* Service layer through a service-client, which is essentially a .service.ts file. These services fire off a CQRS request. There is technically no controllers in this layer for this architecture. The controllers are used if the service had an HTTP API, but it doesnt. These services inject a service-client and call the...
* Repository layer through a repository-client which is essentially a .repository.ts file. These call the DB.

**Reply by arneschreuder** [reply]
> @arneschreuder Does this look correct? I created a resolver using GraphQL and Prisma to query the DB and retrieve all users
> ![Screenshot 2022-03-15 at 13 30 48](https://user-images.githubusercontent.com/73130428/158368776-3b45c335-b0bd-4193-babd-b381b2917618.png)

In this example, the User object should be in `shared/.../users/user.object.ts` and the users resolver should be in `feature/..../resolvers/users/users.resolver.ts`

**Reply by BenjaminOsmers** [reply]
Perfect, I understand now. Thank you.

**Reply by arneschreuder** [reply]
So each of these layers provide:

* API layer: a users.resolver.ts which has a GraphQL API (through code first approach)
* Service layer: a users.service.ts which has an users-service.interface.ts which determine's its API. That service implements CQRS which contains the backend logic. In CQRS, both command handlers queue events and event handlers persist data to the db through a repository instance which is provided next.
* Repository layer: a users.repository.ts which has a users-repository.interface.ts which determine's its API. That service makes use of prisma client to call the DB.

Take note, each of these layers provides an API to the one above it. Those are the "contracts" between entities, services, features that everyone should be working on now.

**Reply by arneschreuder** [reply]
@COS-301/project-managers take note of the conversation above.

**Reply by arneschreuder** [reply]
@COS-301/visual-basic I think your team is architecture right? See this conversation.

**Reply by BenjaminOsmers** [reply]
I'm a bit confused regarding the libraries. When generating library for a feature, should it be generated in /libs/api/{whatever feature name}
And then inside that directory there should be two libraries, one feature library and one shared library. Is this correct?

Or should it be generated in /libs/api/{whatever feature name} and then inside that directory should be api, repository, service directories. Each with a feature and shared library?

**Reply by arneschreuder** [reply]
So the convention is:


* libs -> everything developed is pretty much a micro-lib
* apps -> so this would be either client or api
* feature/domain -> so this would be for example users or authentication
* layer -> so this would be api, service, repository
* library type -> so this would be feature/data-access/util/ui

Shared libraries can be created at each level and then follow:

* shared
* type -> like interfaces, services, etc/
* feature/domain -> like users, authentication
* library type -> mostly ui/data-access/util

Shared libs should be contained to WHO OWNS THAT CODE. These people have the responsibility, that if they want others to consume their service or whatever, they must provide that capability in the shared libs.

**Reply by KyleSmith19091** [reply]
Hi @BenjaminOsmers, I'm part of the Architecture feature(@COS-301/visual-basic) I've started work on the Architecture Wiki which addresses your issue. @arneschreuder explained it nicely, but this is mentioned in the Architecture wiki(https://github.com/COS-301/graduates/wiki/Architecture). Apologies for the poor communication regarding this, but if anything is still unclear just shoot me a message or anyone in my team so we can update the wiki and help you out.

**Reply by BenjaminOsmers** [reply]
Thanks, your wiki helps a lot.

**Reply by arneschreuder** [reply]
See why we have Wikis. So I should've actually just pointed you to that! Nice!

**Reply by Ruben-Brits** [reply]
> Hi @BenjaminOsmers, I'm part of the Architecture feature(@COS-301/visual-basic) I've started work on the Architecture Wiki which addresses your issue. @arneschreuder explained it nicely, but this is mentioned in the Architecture wiki(https://github.com/COS-301/graduates/wiki/Architecture). Apologies for the poor communication regarding this, but if anything is still unclear just shoot me a message or anyone in my team so we can update the wiki and help you out.

@KyleSmith19091  Maybe our PM @DurandtUys can speak to the other PM's just to point their teams towards our Wiki, even though it is still a work in progress

**Reply by BenjaminOsmers** [reply]
@arneschreuder Can you please explain more regarding the repository layer. I don’t understand the necessity of it. 
The API layer contains the resolvers, the Service later contains the Interfaces used by the resolvers. In the Interfaces you would call for instance prisma.findById(). Why is the repository layer necessary?

**Reply by arneschreuder** [reply]
No that is not right:

* API Layer has resolvers. These are basically just components that "resolve" the data they need and present them in a certain way (GraphQL). These resolvers call services.
* Services Layer: These services are the starting point to dispatch CQRS commands, queries and/or events. That is all these services do. In the CQRS pattern, the Command Handlers and Event Handlers will need to access data from the DB. But they are NOT ALLOWED to do so directly. In comes the...
* Repository Layer: The repository provides a layer that is the interface between services and the database. So only the repositories will communicate with the DB and pass that data in a standardised format on to what ever service needs it.

This helps us separate concern and focus on making reusable components.

**Reply by arneschreuder** [reply]
So UI User Profile -> NGXS State -> API Client Service -> GraphQL API App -> UsersResolver.findUnique -> UsersService.findUnique -> UsersRepository.findUnique -> DB

**Reply by BenjaminOsmers** [reply]
@arneschreuder I created an API that would retrieve all student stories from the DB. Does this file structure look correct? (I only added the relevant files in the structure). I assume the story.entity.ts and prisma.service.ts should move to a shared library.
![Screenshot 2022-03-16 at 18 41 40](https://user-images.githubusercontent.com/73130428/158642300-e7d0858b-84b5-4380-aca3-92ce4e0c490c.png)

**Reply by arneschreuder** [reply]
Hi There. Almost...

* prisma.service.ts is a shared service that all data-access libs will use. So it needs to be in libs/api/shared/services/prisma/data-access
* story.entity.ts is this the graphql entity? then it must be in libs/api/stories/api/shared/entities/data-access
* stories.resolver.ts: this is placed fine, but does not call CQRS stuff, this is the service layer's job. The resolver injects storiesService: StoriesService (which is made by the service engineer and shared as a micro-lib for you to consume) and calls storiesService.uniqueStory(....)
* api-stories-api-feature.module.ts: when generating libraries using nx console, a tip is: check the box for "Simple Module Name", then it wont use the entire hierarchy as a filename. Otherwise placed correctly.
* stories.repository.ts: perfect
* there is a missing service in the services layer. This is the StoriesService I was speaking of above. So basically it is a dedicated layer that is responsible for dispatching the CQRS calls. Remember. The API should not unaware of the inner workings of the service layer. So by dispatching Queries and stuff from the resolvers, this defeats the point. Also, we do this split so that services can eventually be run as micro-services and then the API does not care if the service is running on a separate process, because it only cares about what the StoriesService provide.

**Reply by arneschreuder** [reply]
* O yes, and your commands, events and queries can all access the repository. This gets a bit complicated, but we are not going there now. But CQRS is good for when you have disproportionate reads and writes. So it splits the stream of writes and reads into two streams, each possibly with their own database. A read-optimised DB and a write-optimised DB. So a popular architecture and stack using this setup if event-driven architectures, where a command simply mutates a model and pass that model through events. These events write to a eventstore db, the queries can then either read from the aggregate of the event store or from some model snapshots in a normal relational DB for example.

**Reply by BenjaminOsmers** [reply]
Thank you, that clears my last confusion up :)

**Comment by MichaelDuPlessis** [comment]
The API engineers should meet soonish I think to discuss the planning of the schema I believe is the right word for the api's.

**Reply by arneschreuder** [reply]
Agreed. Please use graphql nestjs code first approach

**Comment by MuziwandileTNdlovu** [comment]
A meeting on Monday to get the ball rolling is that okay with all API Engineers at 17:30 does that work with all of us ??

**Reply by BenjaminOsmers** [reply]
Good with me.

**Reply by MichaelDuPlessis** [reply]
That works for me. Where will we be meeting?

**Reply by Grobbies26** [reply]
Thats good with me, I assume it will be on discord?

**Reply by DamianJordaan** [reply]
can we have it Monday morning before the lecture or Tuesday morning before 14:00 or after 17:00

**Reply by BenjaminOsmers** [reply]
Tuesday works with me.

**Reply by RRduToit** [reply]
I can't do before the lecture on monday but the other times work for me

**Reply by MichaelDuPlessis** [reply]
I can do Tuesday after 5.

**Reply by Siphesihle05** [reply]
I'm also good with Tuesday at five.

**Reply by Ying6707** [reply]
I vote for 5pm onwards

**Reply by MuziwandileTNdlovu** [reply]
Cool then Tuesday 6pm

**Reply by takumuguti** [reply]
Tuesday 6pm works for me.

**Reply by MichaelDuPlessis** [reply]
Works for me

**Reply by Dylan-Ku** [reply]
Fine with me

**Reply by Ivy-leet** [reply]
Tuesday 6pm then

**Comment by MuziwandileTNdlovu** [comment]
@arneschreuder  You said for my group(fortran) feature Storage the implementation for an api with graph ql is ugly in terms of running it through docker or firebase , then you went on and said maybe use a rest api does that mean my group has an exception that just having a rest api with nest js ??

**Reply by arneschreuder** [reply]
Yes, if that is the route you choose. You will then be the only feature that has a rest interface. If you do not choose the graphql route. That is up to you.

**Comment by takumuguti** [comment]
@arneschreuder Good day I hope you are well, for my group(basic) you mentioned that I as the API engineer to to build the API for the hosting status page, as well as helping with the deployment of the API app itself. I understand what I need to do for the status page however could I please gain clarity on what is expected of me with regards to helping with the deployment of the API app.

**Reply by arneschreuder** [reply]
Sure, so the api is an app that is going to be run. It contains the api layer, the services layer, the presentation layer. The API must be hosted. Specifically, I asked that the API be bundled into a Docker container so that we can host it like that. You would need to get involved in this effort. I.e. the hosting of the API.

**Reply by takumuguti** [reply]
Thank you for the clarification

**Comment by MichaelDuPlessis** [comment]
@arneschreuder Should a separate API page be created on the wiki describing how to call the different API's and such or should it be put under the API Architecture heading?

**Reply by arneschreuder** [reply]
Sure, go ahead, that might be a good idea. Make a separate page!

**Reply by t3ddy33** [reply]
@arneschreuder Good evening, sir. I'm a tester for the group whose role is that of architecture, and I'm still not sure what exactly is expected of me currently. Could you possibly provide some clarification in this regard?

**Reply by arneschreuder** [reply]
You need to think of testing the status of the architecture and getting a holistic view of strain, capacity, load etc. As a tester you would need to know how you would go about testing the setup of the architecture given some assumptions of traffic, load, scaling requirements etc. You have to help ensure that feature teams provide e2e tests between client and api, these have an architectural scope to it.

**Reply by arneschreuder** [reply]
So if the delivery team comes to you and says there deployment is done, how do you ensure the architecture is actually deployed according to plan. Perhaps go read up on like Terraform and then from that conclude that there probably has to be some Infrastructure-as-Code (IaaC) to be able to do such testing. Terraform for example can be included in CI/CD which then includes some level of architectural testing.

**Comment by MuziwandileTNdlovu** [comment]
@arneschreuder just wanted to clear something , does the schema that the api engineer does end up being the schema used by the data engineer to design their database or are those two separate concerns ? And does the data engineer create the database and design the data base and then the api engineer creates the schema , api and along with that creates and validates the endpoints ??

**Reply by arneschreuder** [reply]
So that is the beauty of layered approaches. Each layer has their own design that fits their layer. The API Engineer designs the API according to what the UI engineer needs. The Service Engineer designs their API according to what the API Engineer needs. And the Data Engineer designs their repository and DB models according to what the Service Engineer needs. HOWEVER, these layers are all in the same "domain". So the concept of an entity, should be agreed upon by ALL LAYERS. Take note, each layer can present that however they want, but the core idea must be agreed upon. So for example, a user entity consists out of firstname, lastname, .... etc. That says nothing of how that must be implemented in each layer. So the repository will work with that data as type, the service layer as a model (class), the repository as a graphql object (class) and the UI as an interface.

**Reply by arneschreuder** [reply]
We have a guest lecture later the year when we talk about integration like this.

**Comment by MichaelDuPlessis** [comment]
@arneschreuder I saw somewhere that we need to have a mock API done by this Sunday. Once we have created this mock API should we submit a pull request with it?

**Reply by arneschreuder** [reply]
I would say you shoul've technically already made PRs with small bits and pieces coming together XD. Each feature has their API engineer and those should each have started with like basic skeleton API stuff, GET IT IN DEVELOP asap :) This is the only way. Small improvement, commit, pr, merge, repeat. Thousands of times!

Do not wait for massive PRs. Get code in. If it is not perfect, we fix it in another PR. Rather break work into small PRs, it makes the flow easier.

**Comment by MichaelDuPlessis** [comment]
@COS-301/api-engineers There is a meeting on Monday 21st of March at 17:00 please let me know if you are unable to attend so I can see if a different meeting time should be planned.

**Comment by MichaelDuPlessis** [comment]
So how I understand everything to work on the API side we will generate a schema using the code first approach and how our API's will be queried is by the UI engineers sending a request to the appropriate endpoint where the request is a JSON object matching the GraphQL schema. Is this line of thinking correct?

**Reply by arneschreuder** [reply]
This sounds right. _I think? hahaha_ I mean, like I think I understand what you are saying. So yes. The API's schema is a cumulative effort of all the API engineers for each feature that attribute their entities and input types to the greater schema. This schema becomes the "docs for the API". You can go and make examples in the wiki. The API-Engineers and UI Engineers can work together or decide under whos scope this falls, but there has to be a client-service implementation of the API that is "just used" in the UI. So like a ApiClientService that is like a wrapper around graphql queries using angular-apollo for example.

**Reply by MichaelDuPlessis** [reply]
Okay great I am understanding and I see that having a controller is not necessary and all you actually need is a module, service and resolver correct?

**Reply by arneschreuder** [reply]
Well, the api layer will be a feature, so it has a feature module that provides its resolvers. The service layer is another feature lib, so that too will have a feature module, but that one provides its service. These are then imported in the shell feature module.

**Comment by MuziwandileTNdlovu** [comment]
![Screenshot (62)](https://user-images.githubusercontent.com/84350184/159144733-5a2e7f07-c088-45ac-8b32-a7254957572d.png)

Is this basically what you would want to see as a mock api ? , and should we make pull requests with our mock api's @arneschreuder

**Reply by arneschreuder** [reply]
Exactly boetie. The actual response does not have to be real, and integrated. Just mocked data returning. Because now the UI engineers can consume this, and go on and would not even notice once you swap out for the real integration with real responses and real data.

**Reply by arneschreuder** [reply]
Obviously the above is a vast oversimplification of what will ultimately be there. But this is the core of the request.

**Reply by MuziwandileTNdlovu** [reply]
@arneschreuder  just to confirm the shell is where we add our mock api as an import ??

**Reply by arneschreuder** [reply]
Kinda yes. The shell is where all the feature modules come together for the server/api app. So each feature has its api feature module + service feature module + repository feature module, and all of those come together in feature module (which imports the api, service and repo modules). All the feature modules are then imported into the shell.

**Comment by takumuguti** [comment]
Has anyone else run into the following error when try to run npm:
```
node:internal/modules/cjs/loader:933
  const err = new Error(message);
              ^

Error: Cannot find module './utils/unsupported.js'
Require stack:
- C:\Users\Taku\AppData\Roaming\npm\node_modules\npm\lib\cli.js
- C:\Users\Taku\AppData\Roaming\npm\node_modules\npm\bin\npm-cli.js
←[90m    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)←[39m
←[90m    at Function.Module._load (node:internal/modules/cjs/loader:778:27)←[39m
←[90m    at Module.require (node:internal/modules/cjs/loader:1005:19)←[39m
←[90m    at require (node:internal/modules/cjs/helpers:102:18)←[39m
    at module.exports (C:\Users\Taku\AppData\Roaming\npm\node_modules\←[4mnpm←[24m\lib\cli.js:14:59)
    at Object.<anonymous> (C:\Users\Taku\AppData\Roaming\npm\node_modules\←[4mnpm←[24m\bin\npm-cli.js:2:25)
←[90m    at Module._compile (node:internal/modules/cjs/loader:1103:14)←[39m
←[90m    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1157:10)←[39m
←[90m    at Module.load (node:internal/modules/cjs/loader:981:32)←[39m
←[90m    at Function.Module._load (node:internal/modules/cjs/loader:822:12)←[39m {
  code: ←[32m'MODULE_NOT_FOUND'←[39m,
  requireStack: [
    ←[32m'C:\\Users\\Taku\\AppData\\Roaming\\npm\\node_modules\\npm\\lib\\cli.js'←[39m,
    ←[32m'C:\\Users\\Taku\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js'←[39m
  ]
}
```
I have searched all over the internet, reinstalled my node.js but nothing seems to be working. The error seems to have started after I upgraded to node.js 16.14.2 (LTS) I also tried to update to node.js 17.7.2 (Current) but nothing seems to be working. I would appreciate any help.

**Reply by MichaelDuPlessis** [reply]
I believe this project is using yarn so that might be your error, I would redownload the repo and try again but use yarn instead

**Reply by takumuguti** [reply]
I will try that, thanks.

**Reply by arneschreuder** [reply]
I agree with @MichaelDuPlessis that is a good place to start.

**Reply by takumuguti** [reply]
I have resolved the issue. The issue was I could not install yarn due to my npm not working however after deleting the npm, and npm cache folders I managed to fix it and install yarn.

**Comment by MichaelDuPlessis** [comment]
What emoji should we use for a mock api commit?

**Reply by arneschreuder** [reply]
From the output of gitmoji cli:


🎨 - :art: - Improve structure / format of the code.
⚡️ - :zap: - Improve performance.
🔥 - :fire: - Remove code or files.
🐛 - :bug: - Fix a bug.
🚑 - :ambulance: - Critical hotfix.
✨ - :sparkles: - Introduce new features.
📝 - :memo: - Add or update documentation.
🚀 - :rocket: - Deploy stuff.
💄 - :lipstick: - Add or update the UI and style files.
🎉 - :tada: - Begin a project.
✅ - :white_check_mark: - Add or update tests.
🔒 - :lock: - Fix security issues.
🔖 - :bookmark: - Release / Version tags.
🚨 - :rotating_light: - Fix compiler / linter warnings.
🚧 - :construction: - Work in progress.
💚 - :green_heart: - Fix CI Build.
⬇️ - :arrow_down: - Downgrade dependencies.
⬆️ - :arrow_up: - Upgrade dependencies.
📌 - :pushpin: - Pin dependencies to specific versions.
👷 - :construction_worker: - Add or update CI build system.
📈 - :chart_with_upwards_trend: - Add or update analytics or track code.
♻️ - :recycle: - Refactor code.
➕ - :heavy_plus_sign: - Add a dependency.
➖ - :heavy_minus_sign: - Remove a dependency.
🔧 - :wrench: - Add or update configuration files.
🔨 - :hammer: - Add or update development scripts.
🌐 - :globe_with_meridians: - Internationalization and localization.
✏️ - :pencil2: - Fix typos.
💩 - :poop: - Write bad code that needs to be improved.
⏪ - :rewind: - Revert changes.
🔀 - :twisted_rightwards_arrows: - Merge branches.
📦 - :package: - Add or update compiled files or packages.
👽 - :alien: - Update code due to external API changes.
🚚 - :truck: - Move or rename resources (e.g.: files, paths, routes).
📄 - :page_facing_up: - Add or update license.
💥 - :boom: - Introduce breaking changes.
🍱 - :bento: - Add or update assets.
♿️ - :wheelchair: - Improve accessibility.
💡 - :bulb: - Add or update comments in source code.
🍻 - :beers: - Write code drunkenly.
💬 - :speech_balloon: - Add or update text and literals.
🗃 - :card_file_box: - Perform database related changes.
🔊 - :loud_sound: - Add or update logs.
🔇 - :mute: - Remove logs.
👥 - :busts_in_silhouette: - Add or update contributor(s).
🚸 - :children_crossing: - Improve user experience / usability.
🏗 - :building_construction: - Make architectural changes.
📱 - :iphone: - Work on responsive design.
🤡 - :clown_face: - Mock things.
🥚 - :egg: - Add or update an easter egg.
🙈 - :see_no_evil: - Add or update a .gitignore file.
📸 - :camera_flash: - Add or update snapshots.
⚗ - :alembic: - Perform experiments.
🔍 - :mag: - Improve SEO.
🏷️ - :label: - Add or update types.
🌱 - :seedling: - Add or update seed files.
🚩 - :triangular_flag_on_post: - Add, update, or remove feature flags.
🥅 - :goal_net: - Catch errors.
💫 - :dizzy: - Add or update animations and transitions.
🗑 - :wastebasket: - Deprecate code that needs to be cleaned up.
🛂 - :passport_control: - Work on code related to authorization, roles and permissions.
🩹 - :adhesive_bandage: - Simple fix for a non-critical issue.
🧐 - :monocle_face: - Data exploration/inspection.

**Reply by arneschreuder** [reply]
I almost always just use ✨ since that is basically any new contribution.

**Reply by arneschreuder** [reply]
🤡 can also be used

**Reply by arneschreuder** [reply]
Don't worry too much about getting this right exactly, this is just meta to help a little bit. 
So the clown face is probably correct. I usually use sparkles, because usually it is more than just a mock, it is like libraries generated and files created etc.

**Comment by BenjaminOsmers** [comment]
@arneschreuder When branching I used git checkout ... instead of git workflow. After my merge, under actions there are "errors" on the workflow runs which state that the workflow was not found. Is this an issue? Or can I delete the workflow runs?

**Reply by arneschreuder** [reply]
Hi There. Can you point me to your fork where this happened? Then I can get better context.

**Reply by BenjaminOsmers** [reply]
It is under COS301/graduates in the action tab.

**Reply by BenjaminOsmers** [reply]
The bottom two.
![Screenshot 2022-03-22 at 21 52 45](https://user-images.githubusercontent.com/73130428/159564340-dbaad3f8-cb80-4b68-90b4-0e020639b96c.png)

**Reply by BenjaminOsmers** [reply]
@arneschreuder Should this be resolved? How can I avoid this in the future?

**Reply by arneschreuder** [reply]
This is resolved now. The pipeline tells you what is wrong in the repo.

**Comment by MuziwandileTNdlovu** [comment]
@COS-301/api-engineers   make sure all your paths use @graduates/.........  notation changing mine now and try make PR again , i think that is what is causing the PR issues

**Reply by takumuguti** [reply]
the following seems to be the issue:
![image](https://user-images.githubusercontent.com/90533439/159652100-6a99e8ab-cd67-4414-9a95-d41bc8eb76e5.png)

**Comment by MuziwandileTNdlovu** [comment]
@arneschreuder after you fix the problem of the PR's could you just explain the mistakes we made so they are not made again is it only that we were not using the @graduates/path notation for importing other libraries in other projects or it is much more than that??

**Reply by arneschreuder** [reply]
Hi There. Sure. So basically this one is a bit weird, since it is a tsconfig thing. I fixed it by adding `!` next to the variable, which tells the compiler that this value will later be initialised. Also, I removed the `= boolean`, because it already knows it is boolean. You can go check the diff of the PR I made to fix this. It should be straight forward. This wasn't a difficult fix.

![image](https://user-images.githubusercontent.com/1528314/159660475-5424eb74-36d5-4dc5-a27c-95ea681febbb.png)

**Reply by MuziwandileTNdlovu** [reply]
Thank you

**Comment by MichaelDuPlessis** [comment]
Just to confirm before we all go in blind but I controller is not necessary and we will be using a resolver instead?

**Reply by arneschreuder** [reply]
No controllers no. Resolvers -> Services. You would have a controller if there was a rest/grpc/kafka/rabbitmq/tpc interface, but not for graphql.

**Comment by BenjaminOsmers** [comment]
@COS-301/pascal your student profiles feature seems to be preventing the API from running. See the screenshot below
![Screenshot 2022-04-12 at 16 35 54](https://user-images.githubusercontent.com/73130428/162987296-a0dddab7-12d2-4dea-907a-ca08afce271a.png)
:

**Reply by BenjaminOsmers** [reply]
The problem is coming from the ApiStudentProfilesEntity class. Which I assume you created?

**Reply by MildogMiller** [reply]
yes sorry i missed that, let me speak to my API engineer

**Reply by BenjaminOsmers** [reply]
Thanks

**Reply by Xander-FP** [reply]
I made the PR with the fix. I will merge when all tests pass.

**Comment by BenjaminOsmers** [comment]
@COS-301/api-engineers Which team deals with logging in a user? I want to know how the session would be storage. More specifically, how would one retrieve the details (either user id or token) of the currently logged in user?

**Reply by Vhulenda88** [reply]
@ricecakesreece

**Reply by MoeMoola3** [reply]
There is a login function implemented in a service file that returns the JSON Web Token

**Reply by BenjaminOsmers** [reply]
Yes but is the userinfo (token, id, etc) then stored in localstorage?
How do other teams access the currently logged in user's info? From my understanding it should be put in localstorage after log in (I could be wrong).

**Reply by MoeMoola3** [reply]
We don't have a frontend yet, but we are trying to find a better way to store it on the front end. They will have access to it on the front end once it is sorted out.

**Reply by BenjaminOsmers** [reply]
Thanks.

**Comment by MichaelDuPlessis** [comment]
I am just struggling to find a way to let the user of my API know that the item they requested for does not exist, I would like to do a custom graphql error message but am struggling to figure out how to do so and currently am just returning an entity with null fields.

**Reply by BenjaminOsmers** [reply]
Do the exception handling in your resolver/service layer. Like in the resolver, if it returns null throw a NotFoundException. Well, that's what I did. Not sure if it is the correct way though.

**Comment by MichaelDuPlessis** [comment]
@arneschreuder My nx console did not generate the spec file, can I just create it myself or what should I do?

**Reply by takumuguti** [reply]
it seems nx does not have an option to generate test files for modules that already exits so I think you should just add it yourself.

**Comment by Vhulenda88** [comment]
![WhatsApp Image 2022-04-18 at 8 30 47 PM](https://user-images.githubusercontent.com/49733146/163860082-1222765d-cdff-45a1-a214-71a923aaf396.jpeg)

@COS-301/api-engineers 
@COS-301/data-engineers 
@COS-301/services-engineers 
does anyone know how to solve this

**Reply by takumuguti** [reply]
in your test make sure to add your resolver as a provider.

**Reply by Vhulenda88** [reply]
we tried that but thanks

**Reply by Vhulenda88** [reply]
@arneschreuder do you mind helping us with this fix, we have tried everything

**Comment by thangeniFT** [comment]
Did you import APIAuthorisationResolver from the right path?

**Reply by Vhulenda88** [reply]
Which Path is it supposed to be ?
